File: frontend/src/services/employeeService.ts
import api from './api';
import { Employee, VacationPeriod, FixedAllocation } from '@/types/planner';

// ========== Types ==========

interface EmployeeDto {
  id: string;
  name: string;
  role: string;
  teamId: string;
  teamName?: string;
  dailyHours: number;
  isActive: boolean;
  createdAt: string;
  vacations: VacationDto[];
  fixedAllocations: FixedAllocationDto[];
}

interface VacationDto {
  id: string;
  startDate: string;
  endDate: string;
}

interface FixedAllocationDto {
  id: string;
  name: string;
  hoursPerMonth: number;
}

interface EmployeeListResponse {
  employees: EmployeeDto[];
  totalCount: number;
}

interface RoleDto {
  value: string;
  label: string;
}

interface RolesListResponse {
  roles: RoleDto[];
}

interface CreateEmployeeRequest {
  name: string;
  role: string;
  departmentId: string;
  dailyHours: number;
  vacations: { startDate: string; endDate: string }[];
  fixedAllocations: { name: string; hoursPerMonth: number }[];
}

interface UpdateEmployeeRequest {
  name: string;
  role: string;
  departmentId: string;
  dailyHours: number;
}

// ========== Helper Functions ==========

const mapDtoToEmployee = (dto: EmployeeDto): Employee => ({
  id: dto.id,
  name: dto.name,
  role: dto.role,
  teamId: dto.teamId,
  dailyHours: dto.dailyHours,
  vacations: dto.vacations.map(v => ({
    id: v.id,
    startDate: new Date(v.startDate),
    endDate: new Date(v.endDate),
  })),
  fixedAllocations: dto.fixedAllocations.map(f => ({
    id: f.id,
    name: f.name,
    hoursPerMonth: f.hoursPerMonth,
  })),
});

// ========== API Service ==========

export const employeeService = {
  async getEmployees(departmentId?: string): Promise<Employee[]> {
    try {
      const params = departmentId ? { departmentId } : {};
      const response = await api.get<EmployeeListResponse>('/api/employees', { params });
      return response.data.employees.map(mapDtoToEmployee);
    } catch (error) {
      console.error('Erro ao buscar funcionários:', error);
      return [];
    }
  },

  async getEmployeeById(id: string): Promise<Employee | null> {
    try {
      const response = await api.get<EmployeeDto>(`/api/employees/${id}`);
      return mapDtoToEmployee(response.data);
    } catch (error) {
      console.error('Erro ao buscar funcionário:', error);
      return null;
    }
  },

  async createEmployee(data: {
    name: string;
    role: string;
    departmentId: string;
    dailyHours: number;
    vacations?: { startDate: Date; endDate: Date }[];
    fixedAllocations?: { name: string; hoursPerMonth: number }[];
  }): Promise<Employee | null> {
    try {
      const request: CreateEmployeeRequest = {
        name: data.name,
        role: data.role,
        departmentId: data.departmentId,
        dailyHours: data.dailyHours,
        vacations: data.vacations?.map(v => ({
          startDate: v.startDate.toISOString(),
          endDate: v.endDate.toISOString(),
        })) || [],
        fixedAllocations: data.fixedAllocations || [],
      };
      const response = await api.post<EmployeeDto>('/api/employees', request);
      return mapDtoToEmployee(response.data);
    } catch (error) {
      console.error('Erro ao criar funcionário:', error);
      return null;
    }
  },

  async updateEmployee(id: string, data: {
    name: string;
    role: string;
    departmentId: string;
    dailyHours: number;
  }): Promise<Employee | null> {
    try {
      const request: UpdateEmployeeRequest = {
        name: data.name,
        role: data.role,
        departmentId: data.departmentId,
        dailyHours: data.dailyHours,
      };
      const response = await api.put<EmployeeDto>(`/api/employees/${id}`, request);
      return mapDtoToEmployee(response.data);
    } catch (error) {
      console.error('Erro ao atualizar funcionário:', error);
      return null;
    }
  },

  async deleteEmployee(id: string): Promise<boolean> {
    try {
      await api.delete(`/api/employees/${id}`);
      return true;
    } catch (error) {
      console.error('Erro ao deletar funcionário:', error);
      return false;
    }
  },

  // Férias
  async addVacation(employeeId: string, startDate: Date, endDate: Date): Promise<VacationPeriod | null> {
    try {
      const response = await api.post<VacationDto>(`/api/employees/${employeeId}/vacations`, {
        startDate: startDate.toISOString(),
        endDate: endDate.toISOString(),
      });
      return {
        id: response.data.id,
        startDate: new Date(response.data.startDate),
        endDate: new Date(response.data.endDate),
      };
    } catch (error) {
      console.error('Erro ao adicionar férias:', error);
      return null;
    }
  },

  async removeVacation(employeeId: string, vacationId: string): Promise<boolean> {
    try {
      await api.delete(`/api/employees/${employeeId}/vacations/${vacationId}`);
      return true;
    } catch (error) {
      console.error('Erro ao remover férias:', error);
      return false;
    }
  },

  // Alocações Fixas
  async addFixedAllocation(employeeId: string, name: string, hoursPerMonth: number): Promise<FixedAllocation | null> {
    try {
      const response = await api.post<FixedAllocationDto>(`/api/employees/${employeeId}/fixed-allocations`, {
        name,
        hoursPerMonth,
      });
      return {
        id: response.data.id,
        name: response.data.name,
        hoursPerMonth: response.data.hoursPerMonth,
      };
    } catch (error) {
      console.error('Erro ao adicionar alocação fixa:', error);
      return null;
    }
  },

  async removeFixedAllocation(employeeId: string, allocationId: string): Promise<boolean> {
    try {
      await api.delete(`/api/employees/${employeeId}/fixed-allocations/${allocationId}`);
      return true;
    } catch (error) {
      console.error('Erro ao remover alocação fixa:', error);
      return false;
    }
  },

  // Cargos
  async getRoles(): Promise<{ value: string; label: string }[]> {
    try {
      const response = await api.get<RolesListResponse>('/api/employees/roles');
      return response.data.roles;
    } catch (error) {
      console.error('Erro ao buscar cargos:', error);
      return [];
    }
  },
};

File: backend-api/src/ResourceFlow.Infrastructure/Services/EmployeeService.cs
using Microsoft.EntityFrameworkCore;
using ResourceFlow.Application.DTOs.Employees;
using ResourceFlow.Application.Interfaces;
using ResourceFlow.Domain.Entities;
using ResourceFlow.Infrastructure.Data;

namespace ResourceFlow.Infrastructure.Services;

public class EmployeeService : IEmployeeService
{
    private readonly ApplicationDbContext _context;

    public EmployeeService(ApplicationDbContext context)
    {
        _context = context;
    }

    public async Task<EmployeeListResponse> GetEmployeesAsync(Guid? departmentId = null)
    {
        var query = _context.Employees
            .Include(e => e.Department)
            .Include(e => e.Vacations)
            .Include(e => e.FixedAllocations)
            .Where(e => e.IsActive);

        if (departmentId.HasValue)
        {
            query = query.Where(e => e.DepartmentId == departmentId.Value);
        }

        var employees = await query.OrderBy(e => e.Name).ToListAsync();

        return new EmployeeListResponse
        {
            Employees = employees.Select(MapToDto).ToList(),
            TotalCount = employees.Count
        };
    }

    public async Task<EmployeeDto?> GetEmployeeByIdAsync(Guid id)
    {
        var employee = await _context.Employees
            .Include(e => e.Department)
            .Include(e => e.Vacations)
            .Include(e => e.FixedAllocations)
            .FirstOrDefaultAsync(e => e.Id == id);

        return employee != null ? MapToDto(employee) : null;
    }

    public async Task<EmployeeDto> CreateEmployeeAsync(CreateEmployeeRequest request)
    {
        var employee = new Employee
        {
            Id = Guid.NewGuid(),
            Name = request.Name,
            Role = request.Role,
            DepartmentId = request.DepartmentId,
            DailyHours = request.DailyHours,
            IsActive = true,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };

        _context.Employees.Add(employee);

        // Adicionar férias
        foreach (var vacation in request.Vacations)
        {
            _context.EmployeeVacations.Add(new EmployeeVacation
            {
                Id = Guid.NewGuid(),
                EmployeeId = employee.Id,
                StartDate = vacation.StartDate,
                EndDate = vacation.EndDate
            });
        }

        // Adicionar alocações fixas
        foreach (var allocation in request.FixedAllocations)
        {
            _context.EmployeeFixedAllocations.Add(new EmployeeFixedAllocation
            {
                Id = Guid.NewGuid(),
                EmployeeId = employee.Id,
                Name = allocation.Name,
                HoursPerMonth = allocation.HoursPerMonth
            });
        }

        await _context.SaveChangesAsync();

        return (await GetEmployeeByIdAsync(employee.Id))!;
    }

    public async Task<EmployeeDto?> UpdateEmployeeAsync(Guid id, UpdateEmployeeRequest request)
    {
        var employee = await _context.Employees.FindAsync(id);
        if (employee == null) return null;

        employee.Name = request.Name;
        employee.Role = request.Role;
        employee.DepartmentId = request.DepartmentId;
        employee.DailyHours = request.DailyHours;
        employee.UpdatedAt = DateTime.UtcNow;

        await _context.SaveChangesAsync();

        return await GetEmployeeByIdAsync(id);
    }

    public async Task<bool> DeleteEmployeeAsync(Guid id)
    {
        var employee = await _context.Employees.FindAsync(id);
        if (employee == null) return false;

        employee.IsActive = false;
        employee.UpdatedAt = DateTime.UtcNow;

        await _context.SaveChangesAsync();
        return true;
    }

    // Férias
    public async Task<VacationDto> AddVacationAsync(Guid employeeId, CreateVacationRequest request)
    {
        var vacation = new EmployeeVacation
        {
            Id = Guid.NewGuid(),
            EmployeeId = employeeId,
            StartDate = request.StartDate,
            EndDate = request.EndDate
        };

        _context.EmployeeVacations.Add(vacation);
        await _context.SaveChangesAsync();

        return new VacationDto
        {
            Id = vacation.Id,
            StartDate = vacation.StartDate,
            EndDate = vacation.EndDate
        };
    }

    public async Task<bool> RemoveVacationAsync(Guid employeeId, Guid vacationId)
    {
        var vacation = await _context.EmployeeVacations
            .FirstOrDefaultAsync(v => v.Id == vacationId && v.EmployeeId == employeeId);

        if (vacation == null) return false;

        _context.EmployeeVacations.Remove(vacation);
        await _context.SaveChangesAsync();
        return true;
    }

    // Alocações Fixas
    public async Task<FixedAllocationDto> AddFixedAllocationAsync(Guid employeeId, CreateFixedAllocationRequest request)
    {
        var allocation = new EmployeeFixedAllocation
        {
            Id = Guid.NewGuid(),
            EmployeeId = employeeId,
            Name = request.Name,
            HoursPerMonth = request.HoursPerMonth
        };

        _context.EmployeeFixedAllocations.Add(allocation);
        await _context.SaveChangesAsync();

        return new FixedAllocationDto
        {
            Id = allocation.Id,
            Name = allocation.Name,
            HoursPerMonth = allocation.HoursPerMonth
        };
    }

    public async Task<bool> RemoveFixedAllocationAsync(Guid employeeId, Guid allocationId)
    {
        var allocation = await _context.EmployeeFixedAllocations
            .FirstOrDefaultAsync(a => a.Id == allocationId && a.EmployeeId == employeeId);

        if (allocation == null) return false;

        _context.EmployeeFixedAllocations.Remove(allocation);
        await _context.SaveChangesAsync();
        return true;
    }

    // Cargos
    public Task<RolesListResponse> GetRolesAsync()
    {
        var roles = EmployeeRoles.All.Select(r => new RoleDto
        {
            Value = r,
            Label = r
        }).ToList();

        return Task.FromResult(new RolesListResponse { Roles = roles });
    }

    private static EmployeeDto MapToDto(Employee employee)
    {
        return new EmployeeDto
        {
            Id = employee.Id,
            Name = employee.Name,
            Role = employee.Role,
            TeamId = employee.DepartmentId,
            TeamName = employee.Department?.Name,
            DailyHours = employee.DailyHours,
            IsActive = employee.IsActive,
            CreatedAt = employee.CreatedAt,
            Vacations = employee.Vacations.Select(v => new VacationDto
            {
                Id = v.Id,
                StartDate = v.StartDate,
                EndDate = v.EndDate
            }).ToList(),
            FixedAllocations = employee.FixedAllocations.Select(f => new FixedAllocationDto
            {
                Id = f.Id,
                Name = f.Name,
                HoursPerMonth = f.HoursPerMonth
            }).ToList()
        };
    }
}
